tests/pubsub/testgossipmembership.nim

```diff
@@ -0,0 +1,240 @@
+
+    subscribeToTopics(gossipSub, @[topic])
+
+    unsubscribeFromTopics(gossipSub, @[topic])
+
+    check topic notin gossipSub.topics
+    check topic notin gossipSub.mesh
+    check topic in gossipSub.gossipsub
+
+    await allFuturesThrowing(conns.mapIt(it.close()))
+    await gossipSub.switch.stop()
+
+  # Test subscribing and unsubscribing multiple topics
+  asyncTest "handle SUBSCRIBE and UNSUBSCRIBE multiple topics":
+    let topics = ["topic1", "topic2", "topic3"].toSeq()
+    let (gossipSub, conns) = setupGossipSub(topics, 5)
+
+    subscribeToTopics(gossipSub, topics)
+
+    check gossipSub.topics.len == 3
+    for topic in topics:
+      check gossipSub.gossipsub[topic].len() == 5
+
+    unsubscribeFromTopics(gossipSub, topics)
+
+    for topic in topics:
+      check topic notin gossipSub.topics
+      check topic notin gossipSub.mesh
+      check topic in gossipSub.gossipsub
+
+    await allFuturesThrowing(conns.mapIt(it.close()))
+    await gossipSub.switch.stop()
+
+  # Test ensuring that the number of subscriptions does not exceed the limit set in the GossipSub parameters
+  asyncTest "subscription limit test":
+    let topicCount = 15
+    let gossipSubParams = 10
+    let topicNames = toSeq(mapIt(0 .. topicCount - 1, "topic" & $it))
+
+    let (gossipSub, conns) = setupGossipSub(topicNames, 0)
+
+    gossipSub.topicsHigh = gossipSubParams
+
+    for topic in topicNames:
+      if gossipSub.topics.len < gossipSub.topicsHigh:
+        gossipSub.subscribe(
+          topic,
+          proc(topic: string, data: seq[byte]): Future[void] {.async.} =
+            discard
+          ,
+        )
+      else:
+        check gossipSub.topics.len == gossipSub.topicsHigh
+
+    check gossipSub.topics.len <= gossipSub.topicsHigh
+    check gossipSub.topics.len == gossipSub.topicsHigh
+
+    await allFuturesThrowing(conns.mapIt(it.close()))
+    await gossipSub.switch.stop()
+
+  # Test for verifying peers joining a topic using `JOIN(topic)`
+  asyncTest "handle JOIN topic and mesh is updated":
+    let topic = "test-join-topic"
+
+    let (gossipSub, conns) = setupGossipSub(topic, 5)
+
+    commonSubscribe(@[gossipSub], topic, voidTopicHandler)
+
+    check gossipSub.mesh[topic].len == 5
+    check gossipSub.topics.contains(topic)
+
+    await allFuturesThrowing(conns.mapIt(it.close()))
+    await gossipSub.switch.stop()
+
+  # Test the behavior when multiple peers join and leave a topic simultaneously.
+  asyncTest "multiple peers join and leave topic simultaneously":
+    let
+      numberOfNodes = 6
+      topic = "foobar"
+      nodes = generateNodes(numberOfNodes, gossip = true)
+      nodesFut = await allFinished(nodes.mapIt(it.switch.start()))
+
+    await subscribeNodes(nodes)
+    for node in nodes:
+      node.subscribe(topic, voidTopicHandler)
+
+    await sleepAsync(2 * DURATION_TIMEOUT)
+
+    for i in 0 ..< numberOfNodes:
+      let currentGossip = GossipSub(nodes[i])
+      check currentGossip.gossipsub.hasKey(topic)
```

Comment:
If you're checking the state, which is a great idea, I'd suggest checking it completely. I mean, checking all relevant attributes to this test: `gossipsub`, `mesh`. Maybe `topics` as well, but this last one is not that important.